### ISSUE TITLE
[staticcheck,SA5011] false positives in tests

### ISSUE THREAD

**Issue by kolyshkin** [golangci/golangci-lint#5979]
### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/usage/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

Official GitHub Action

### Description of the problem

In our repo (https://github.com/opencontainers/runc) we use golangci-lint action, and from time to time I see false positives like this:

```
 Error: libcontainer/capabilities/capabilities_linux_test.go:53:5: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
  	if l == nil {
  	   ^
  Error: libcontainer/capabilities/capabilities_linux_test.go:56:7: SA5011: possible nil pointer dereference (staticcheck)
  	if l.Level != expectedLogs.Level {
  	     ^
  Error: libcontainer/configs/tocpuset_test.go:81:7: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
  			if out == nil {
  			   ^
  Error: libcontainer/configs/tocpuset_test.go:84:7: SA5011: possible nil pointer dereference (staticcheck)
  			if *out != *tc.out {
  			   ^
  Error: libcontainer/system/kernelversion/kernel_linux_test.go:32:5: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
  	if version == nil {
  	   ^
  Error: libcontainer/system/kernelversion/kernel_linux_test.go:35:13: SA5011: possible nil pointer dereference (staticcheck)
  	if version.Kernel == 0 {
  	           ^
  6 issues:
  * staticcheck: 6
```

In all 3 cases, the  code looks like this:

```go
        if l == nil {
                t.Fatal("expected a warning, but got none")
        }
        if l.Level != expectedLogs.Level {
                t.Errorf("expected %q, got %q", expectedLogs.Level, l.Level)
        }
```

NOTE
 * this does not happens always, but occasionally, from time to time;
 * re-running the same job does not help;
 * clearing golangci-cache always helps.

My gut feeling this is something related to the cache.

Here is a run which uses cache (with false positives): https://github.com/opencontainers/runc/actions/runs/16771007365/job/47485871649

Here's a second run (no changes): https://github.com/opencontainers/runc/actions/runs/16771007365/job/47536430979

Here's a third run, with the cache (golangci-lint.cache-Linux-2900-07b6deaeaf77c89136c312b904939c6615b804c9') manually removed, (no more false positives): https://github.com/opencontainers/runc/actions/runs/16771007365/job/47536777801

What puzzles me is the last job still says:

```
 Cache hit for: golangci-lint.cache-Linux-2900-07b6deaeaf77c89136c312b904939c6615b804c9
  Received 641357 of 641357 (100.0%), 9.7 MBs/sec
  Cache Size: ~1 MB (641357 B)
  /usr/bin/tar -xf /home/runner/work/_temp/f540f2d7-e2d1-4b51-9da4-0209e70389ca/cache.tzst -P -C /home/runner/work/runc/runc --use-compress-program unzstd
  Cache restored successfully
  Restored cache for golangci-lint from key 'golangci-lint.cache-Linux-2900-07b6deaeaf77c89136c312b904939c6615b804c9' in 404ms
```

although I manually removed this very cache just before re-running the job, so something might be wrong in github actions department, too.

### Version of golangci-lint

From the gha log:
<details>

```
  Finding needed golangci-lint version...
  Requested golangci-lint 'v2.3', using 'v2.3.1', calculation took 56ms
  Installation mode: binary
  Installing golangci-lint binary v2.3.1...
  Downloading binary https://github.com/golangci/golangci-lint/releases/download/v2.3.1/golangci-lint-2.3.1-linux-amd64.tar.gz ...
  /usr/bin/tar xz --overwrite --warning=no-unknown-keyword --overwrite -C /home/runner -f /home/runner/work/_temp/f2018b3e-500b-44ad-a3d6-5109026baf62
  Installed golangci-lint into /home/runner/golangci-lint-2.3.1-linux-amd64/golangci-lint in 435ms
  Prepared env in 896ms
```
</details>

### Configuration

See https://github.com/opencontainers/runc/blob/127e8e68d335aa7b908c7822fa985629d8fbe81d/.golangci.yml

### Go environment

From the job logs:
<details>

```
Run actions/setup-go@v5
  with:
    go-version: 1.24
    check-latest: false
    token: ***
    cache: true
  env:
    GO_VERSION: 1.24
Setup go version spec 1.24
Found in cache @ /opt/hostedtoolcache/go/1.24.5/x64
Added go to the path
Successfully set up Go version 1.24
/opt/hostedtoolcache/go/1.24.5/x64/bin/go env GOMODCACHE
/opt/hostedtoolcache/go/1.24.5/x64/bin/go env GOCACHE
/home/runner/go/pkg/mod
/home/runner/.cache/go-build
Cache hit for: setup-go-Linux-x64-ubuntu24-go-1.24.5-2a435ec440cb466e7517b049c14e40c85fd4691bea511023e036a40f9c346cf3
Received 18690705 of 18690705 (100.0%), 87.4 MBs/sec
Cache Size: ~18 MB (18690705 B)
/usr/bin/tar -xf /home/runner/work/_temp/1a117bb1-dab9-4a76-899d-e2661d1695d1/cache.tzst -P -C /home/runner/work/runc/runc --use-compress-program unzstd
Cache restored successfully
Cache restored from key: setup-go-Linux-x64-ubuntu24-go-1.24.5-2a435ec440cb466e7517b049c14e40c85fd4691bea511023e036a40f9c346cf3
go version go1.24.5 linux/amd64

go env
  AR='ar'
  CC='gcc'
  CGO_CFLAGS='-O2 -g'
  CGO_CPPFLAGS=''
  CGO_CXXFLAGS='-O2 -g'
  CGO_ENABLED='1'
  CGO_FFLAGS='-O2 -g'
  CGO_LDFLAGS='-O2 -g'
  CXX='g++'
  GCCGO='gccgo'
  GO111MODULE=''
  GOAMD64='v1'
  GOARCH='amd64'
  GOAUTH='netrc'
  GOBIN=''
  GOCACHE='/home/runner/.cache/go-build'
  GOCACHEPROG=''
  GODEBUG=''
  GOENV='/home/runner/.config/go/env'
  GOEXE=''
  GOEXPERIMENT=''
  GOFIPS140='off'
  GOFLAGS=''
  GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build444757353=/tmp/go-build -gno-record-gcc-switches'
  GOHOSTARCH='amd64'
  GOHOSTOS='linux'
  GOINSECURE=''
  GOMOD='/home/runner/work/runc/runc/go.mod'
  GOMODCACHE='/home/runner/go/pkg/mod'
  GONOPROXY=''
  GONOSUMDB=''
  GOOS='linux'
  GOPATH='/home/runner/go'
  GOPRIVATE=''
  GOPROXY='https://proxy.golang.org,direct/'
  GOROOT='/opt/hostedtoolcache/go/1.24.5/x64'
  GOSUMDB='sum.golang.org'
  GOTELEMETRY='local'
  GOTELEMETRYDIR='/home/runner/.config/go/telemetry'
  GOTMPDIR=''
  GOTOOLCHAIN='auto'
  GOTOOLDIR='/opt/hostedtoolcache/go/1.24.5/x64/pkg/tool/linux_amd64'
  GOVCS=''
  GOVERSION='go1.24.5'
  GOWORK=''
  PKG_CONFIG='pkg-config'
```
</details>

### Verbose output of running

I can't reproduce it locally.

### A minimal reproducible example or link to a public repository

See above.

### Validation

- [x] Yes, I've included all information above (version, config, etc.).

### Supporter

- [ ] I am a sponsor/backer through [GitHub](https://github.com/sponsors/golangci/) or [OpenCollective](https://opencollective.com/golangci-lint)

**Comment by ldez** [comment]
Hello,

I think this is the same root cause as #3228: cache and facts.

**Comment by kolyshkin** [comment]
This has been happening for the last few months, here's the earliest run I was able to find today, it is from May 28 and used golangci-lint v2.1.6: https://github.com/opencontainers/runc/actions/runs/15309263437/job/43069694659

It might be happened earlier than that but the older logs are no longer available.

**Comment by kolyshkin** [comment]
> Hello,
> 
> I think this is the same root cause as [#3228](https://github.com/golangci/golangci-lint/issues/3228): cache and facts.

Might be; I've seen some random false positives from nolintlint, too. Here's one (from the job mentioned in the previous comment):

```
 Error: libcontainer/criu_linux.go:186:61: directive `//nolint:staticcheck // SA1019: strings.Title is deprecated` is unused for linter "staticcheck" (nolintlint)
  	return "extRoot" + strings.Title(configs.NsName(t)) + "NS" //nolint:staticcheck // SA1019: strings.Title is deprecated
  	                                                           ^
```

@ldez If there's anything I can do to help, let me know.

**Comment by kolyshkin** [comment]
I just saw this issue locally, so no GHA is to blame:

```console
[kir@kir-tp1 go-systemd]$ golangci-lint version
golangci-lint has version 2.3.0 built with go1.24.4 from 364a4bbe on 2025-07-21T14:49:16Z
[kir@kir-tp1 go-systemd]$ golangci-lint  run --enable-only staticcheck ./...
0 issues.
[kir@kir-tp1 go-systemd]$ golangci-lint  run ./...
.....
dbus/methods_test.go:711:5: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
	if found == nil {
	   ^
dbus/methods_test.go:715:22: SA5011: possible nil pointer dereference (staticcheck)
	if isJobStatusEmpty(*found) {
	                    ^
```

The code in question is [here](https://github.com/coreos/go-systemd/blob/66963a17145e2e9853c4c52d2f4ac68748baa7f5/dbus/methods_test.go#L710-L717):
```go
	found := getJobStatusIfExists(jobs, service)
	if found == nil {
		t.Fatalf("%s job not found in list", service)
	}

	if isJobStatusEmpty(*found) {
		t.Fatalf("empty %s job found in list", service)
	}
```

This is 100% reproducible so I guess I can debug it if needed. Some add'l info:
```console
[kir@kir-tp1 go-systemd]$ go version
go version go1.24.5 linux/amd64
[kir@kir-tp1 go-systemd]$ golangci-lint  cache status
Dir: /home/kir/.cache/golangci-lint
Size: 1.5MiB
```

**Comment by stevehipwell** [comment]
I'm seeing this locally too.

**Comment by Antonboom** [comment]
```go
func assertFuncFirstParamName(t *testing.T, ft *ast.FuncType, name string) {
	t.Helper()

	if ft == nil {
		t.FailNow()
	}

	if ft.Params.List[0].Names[0].Name != name {
		t.FailNow()
	}
}
```

```
Lint...
pkg/analyzer/func_type_stack_test.go:46:5: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
        if ft == nil {
           ^
pkg/analyzer/func_type_stack_test.go:50:8: SA5011: possible nil pointer dereference (staticcheck)
        if ft.Params.List[0].Names[0].Name != name {
```
